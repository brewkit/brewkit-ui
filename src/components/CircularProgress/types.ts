import { CircularProgressProps } from '@material-ui/core/CircularProgress';
import { Theme } from '@material-ui/core/styles/createMuiTheme';


/**
 * Declare any theme variables we want to make available.
 */
declare module '@material-ui/core/styles/createMuiTheme' {
    interface Theme {
        BkCircularProgress: {
            props: {
                layered: boolean,
            },
        },
    }
    // allow configuration using `createMuiTheme`
    interface ThemeOptions {
        BkCircularProgress: {
            props: {
                layered?: boolean,
            },
        },
    }
}


export type Props = CircularProgressProps & {

    /**
     * adds a background circle
     */
    layered?: boolean,

};
