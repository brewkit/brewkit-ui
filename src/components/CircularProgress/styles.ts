import { makeStyles } from '@material-ui/core/styles';


const useStyles = makeStyles({

    root: {
        display: 'inline-block',
        position: 'relative',
        '& > *:first-child': {
            position: 'absolute',
            opacity: 0.25,
        },
    },

});


export default useStyles;
