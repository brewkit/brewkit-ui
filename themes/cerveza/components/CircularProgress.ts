import { size } from '../variables';


const Badge = {

    props: {

        MuiCircularProgress: {
            thickness: 3,
            size: size.xxlarge,
        },

    },

    BkCircularProgress: {
        props: {
            layered: true,
        },
    },

};


export default Badge;
