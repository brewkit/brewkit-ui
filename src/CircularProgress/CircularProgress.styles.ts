import { createStyles, makeStyles } from '@material-ui/core/styles';

declare module '@material-ui/core/styles/createTheme' {
    interface Theme {
        BkCircularProgress: {
            props: {
                layered: boolean;
            };
        };
    }
    interface ThemeOptions {
        BkCircularProgress: {
            props: {
                layered?: boolean;
            };
        };
    }
}

const useStyles = makeStyles({
    root: {
        display: 'inline-block',
        position: 'relative',
        '& > *:first-child': {
            position: 'absolute',
            opacity: 0.25,
        },
    },
});

export default useStyles;
