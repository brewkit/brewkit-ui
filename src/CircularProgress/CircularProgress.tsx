import * as React from 'react';
import MUICircularProgess, {
    CircularProgressProps as MUICircularProgressProps,
} from '@material-ui/core/CircularProgress';
import { useTheme } from '@material-ui/core/styles';
import useStyles from './CircularProgress.styles';
import { merge } from 'lodash';

export type BkCircularProgressProps = MUICircularProgressProps & {
    /**
     * adds a background circle
     *
     * @default true
     */
    layered?: boolean;
};

export const CircularProgress = React.forwardRef(
    (
        props: BkCircularProgressProps,
        ref: React.Ref<HTMLDivElement>,
    ): React.ReactElement => {
        const { sizes } = useTheme();
        const {
            classes: userClasses,
            thickness = 3,
            size = sizes.xxlarge,
            layered = true,
            ...otherProps
        } = props;
        const classes = merge(useStyles(), userClasses);
        const sharedProps = {
            thickness,
            size,
            ...otherProps,
        };

        /**
         * if layered=true, we layer another CircularProgress in the background with a lower opacity
         */
        if (layered)
            return (
                <div className={classes.root}>
                    <MUICircularProgess
                        {...sharedProps}
                        value={100}
                        variant="determinate"
                    />
                    <MUICircularProgess {...sharedProps} ref={ref} />
                </div>
            );

        return <MUICircularProgess {...sharedProps} ref={ref} />;
    },
);

CircularProgress.displayName = 'BkCircularProgress';

export default CircularProgress;
