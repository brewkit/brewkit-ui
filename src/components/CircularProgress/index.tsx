import * as React from 'react';
import MuiCircularProgress from '@material-ui/core/CircularProgress';
import { useTheme } from '@material-ui/core/styles';
import useStyles from './styles';
import { Props } from './types';


/**
 * Progress indicators commonly known as spinners, express an unspecified wait time or display the length of a process.
 * The animation works with CSS, not JavaScript.
 *
 * [Material-UI Docs](https://material-ui.com/components/progress/)
 *
 * ## Differences in Brewkit:
 *
 * - Adds a `layered` property that can be used to render an always visible circular background. The default value of
 * this prop can be set using `BkCircularProgress.props.layered` in your theme.
 */
export const CircularProgress = React.forwardRef(({
    layered,
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => {


    const theme = useTheme();
    const classes = useStyles();
    const isLayered = theme?.BkCircularProgress?.props?.layered || layered;


    /**
     * if layered=true, we layer another CircularProgress in the background with a lower opacity
     */
    if (isLayered) return (
        <div className={classes.root} ref={ref}>
            <MuiCircularProgress {...otherProps} value={100} variant="static" />
            <MuiCircularProgress {...otherProps} ref={ref} />
        </div>
    );


    return (
        <MuiCircularProgress {...otherProps} ref={ref} />
    );


});


CircularProgress.displayName = 'CircularProgress';


export default CircularProgress;
