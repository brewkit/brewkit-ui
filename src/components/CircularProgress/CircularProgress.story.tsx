
import * as React from 'react';
import { Story } from '@storybook/react';

import CircularProgress from './CircularProgress';
import { Props as CircularProgressProps } from './CircularProgress.types';

export default {
    component: CircularProgress,
    title: 'Feedback/CircularProgress',
};

const defaultArgs = {
    disableShrink: false,
    layered: false,
    size: '2rem',
    thickness: 4,
    value: 0,
};

export const Sandbox: Story<CircularProgressProps> = (args) => (
    <CircularProgress {...args} />
);

Sandbox.args = {
    ...defaultArgs,
    color: 'primary',
    variant: 'indeterminate',
};

export const Indeterminate: Story<CircularProgressProps> = (args) => (
    <div style={{ display: 'grid', gridAutoFlow: 'column', gridGap: '1rem' }}>
        <CircularProgress {...args} />
        <CircularProgress {...args} color="primary" />
        <CircularProgress {...args} color="secondary" />
        <CircularProgress {...args} color="inherit" />
    </div>
);

Indeterminate.args = {
    ...defaultArgs,
    variant: 'indeterminate',
};

export const Determinate: Story<CircularProgressProps> = (args) => {
    const [completed, setCompleted] = React.useState(0);

    React.useEffect(() => {
        const progress = (): void => {
            setCompleted((prevCompleted) => prevCompleted >= 100 ? 0 : prevCompleted + 10);
        };

        const timer = setInterval(progress, 1000);

        return (): void => {
            clearInterval(timer);
        };
    }, []);

    return (
        <div
            style={{
                display: 'grid',
                gridAutoFlow: 'column',
                gridGap: '1rem',
            }}>
            <CircularProgress {...args} value={25} variant="determinate" />
            <CircularProgress {...args} value={50} variant="determinate" />
            <CircularProgress {...args} value={75} variant="determinate" />
            <CircularProgress {...args} value={100} variant="determinate" />
            <CircularProgress
                {...args}
                value={completed}
                variant="determinate"
            />
        </div>
    );
};

Determinate.args = {
    ...defaultArgs,
};

export const Layered: Story<CircularProgressProps> = (args) => (
    <div style={{ display: 'grid', gridAutoFlow: 'column', gridGap: '1rem' }}>
        <CircularProgress {...args} layered />
        <CircularProgress {...args} color="primary" layered />
        <CircularProgress {...args} color="secondary" layered />
        <CircularProgress {...args} color="inherit" layered />
    </div>
);

Layered.args = {
    ...defaultArgs,
};
