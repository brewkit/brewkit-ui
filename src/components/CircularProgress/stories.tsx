import * as React from 'react';
import { text, select, boolean, number } from '@storybook/addon-knobs';
import CircularProgress from '@components/CircularProgress';


export default {
    component: CircularProgress,
    parameters: {
        // componentSubtitle: <Flag color="success">Stable</Flag>,
    },
    title: 'Feedback/CircularProgress',
};


export const Sandbox = (): React.ReactElement => {

    const color = select('color', ['primary', 'secondary', 'inherit'], 'primary');
    const variant = select('variant', ['determinate', 'indeterminate', 'static'], 'indeterminate');
    const disableShrink = boolean('disableShrink', false);
    const layered = boolean('layered', true);
    const size = text('size', '2rem');
    const thickness = number('thickness', 4);
    const value = number('value', 0);

    return (
        <CircularProgress
            color={color}
            variant={variant}
            disableShrink={disableShrink}
            size={size}
            thickness={thickness}
            value={value}
            layered={layered}
        />
    );

};


export const Indeterminate = (): React.ReactElement => (
    <div style={{ display: 'grid', gridAutoFlow: 'column', gridGap: '1rem' }}>

        <CircularProgress layered={false} />
        <CircularProgress layered={false} color="primary" />
        <CircularProgress layered={false} color="secondary" />
        <CircularProgress layered={false} color="inherit" />

    </div>
);


export const Determinate = (): React.ReactElement => {


    const [completed, setCompleted] = React.useState(0);


    React.useEffect(() => {
        function progress() {
            setCompleted(prevCompleted => (prevCompleted >= 100 ? 0 : prevCompleted + 10));
        }

        const timer = setInterval(progress, 1000);

        return () => {
            clearInterval(timer);
        };
    }, []);


    return (
        <div style={{ display: 'grid', gridAutoFlow: 'column', gridGap: '1rem' }}>

            <CircularProgress variant="static" value={25} />
            <CircularProgress variant="static" value={50} />
            <CircularProgress variant="static" value={75} />
            <CircularProgress variant="static" value={100} />
            <CircularProgress variant="static" value={completed} />

        </div>
    )

};


export const Layered = (): React.ReactElement => (
    <div style={{ display: 'grid', gridAutoFlow: 'column', gridGap: '1rem' }}>

        <CircularProgress layered />
        <CircularProgress layered color="primary" />
        <CircularProgress layered color="secondary" />
        <CircularProgress layered color="inherit" />

    </div>
);
