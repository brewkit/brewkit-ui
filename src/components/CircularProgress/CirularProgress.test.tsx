// @ts-nocheck
import * as React from 'react';
import { create } from 'react-test-renderer';
import CircularProgress from '.';

it('Render CircularProgress component', () => {
    const comp = create(
        <CircularProgress />
    ).toJSON();

    expect(comp).toMatchSnapshot();
});

it('Render CircularProgress component with layered prop', () => {
    const comp = create(
        <CircularProgress layered />
    ).toJSON();

    expect(comp).toMatchSnapshot();
});
