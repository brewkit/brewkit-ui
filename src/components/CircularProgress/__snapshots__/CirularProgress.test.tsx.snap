// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render CircularProgress component 1`] = `
<div
  className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
  role="progressbar"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
>
  <svg
    className="MuiCircularProgress-svg"
    viewBox="22 22 44 44"
  >
    <circle
      className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
      cx={44}
      cy={44}
      fill="none"
      r={20.2}
      strokeWidth={3.6}
      style={Object {}}
    />
  </svg>
</div>
`;

exports[`Render CircularProgress component with layered prop 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    aria-valuenow={100}
    className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-determinate"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "transform": "rotate(-90deg)",
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={
          Object {
            "strokeDasharray": "126.920",
            "strokeDashoffset": "0.000px",
          }
        }
      />
    </svg>
  </div>
  <div
    className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </div>
</div>
`;
